src/engine/sync/abletonlink.cpp

```diff
@@ -0,0 +1,248 @@
+    // The callback is the only entity, which access m_pNumLinkPeers
+    m_pLink->setNumPeersCallback([this](std::size_t numPeers) {
+        m_pNumLinkPeers->forceSet(numPeers);
+    });
+
+    m_pLink->enable(false);
+    m_pLink->enableStartStopSync(false);
+
+    nonAudioPrint();
+    audioSafePrint();
+
+    initTestTimer(1000, true);
+}
+
+AbletonLink::~AbletonLink() {
+    m_pLink.reset();
+    delete m_pLinkButton;
+}
+
+void AbletonLink::slotControlSyncEnabled(double value) {
+    qDebug() << "AbletonLink::slotControlSyncEnabled" << value;
+    if (value > 0) {
+        m_pLink->enable(true);
+        m_pLink->enableStartStopSync(true);
+    } else {
+        m_pLink->enable(false);
+        m_pLink->enableStartStopSync(false);
+    }
+}
+
+void AbletonLink::setSyncMode(SyncMode mode) {
+    m_mode = mode;
+}
+
+void AbletonLink::notifyUniquePlaying() {
+}
+
+void AbletonLink::requestSync() {
+}
+
+SyncMode AbletonLink::getSyncMode() const {
+    return m_mode;
+}
+
+bool AbletonLink::isPlaying() const {
+    if (!m_pLink->isEnabled()) {
+        return false;
+    }
+    if (m_pLink->numPeers() < 1) {
+        return false;
+    }
+
+    ableton::Link::SessionState sessionState = m_pLink->captureAudioSessionState();
+    return sessionState.isPlaying();
+}
+bool AbletonLink::isAudible() const {
+    return false;
+}
+bool AbletonLink::isQuantized() const {
+    return true;
+}
+
+mixxx::Bpm AbletonLink::getBpm() const {
+    return getBaseBpm();
+}
+
+double AbletonLink::getBeatDistance() const {
+    ableton::Link::SessionState sessionState = m_pLink->captureAudioSessionState();
+    auto beats = sessionState.beatAtTime(getHostTimeAtSpeaker(getHostTime()), getQuantum());
+    return std::fmod(beats, 1.);
+}
+
+mixxx::Bpm AbletonLink::getBaseBpm() const {
+    ableton::Link::SessionState sessionState = m_pLink->captureAudioSessionState();
+    mixxx::Bpm tempo(sessionState.tempo());
+    return tempo;
+}
+
+void AbletonLink::readAudioBufferMicros() {
+    if ((m_pEngineSync != nullptr) &&
+            (m_pEngineSync->getLeaderChannel() != nullptr) &&
+            (m_pEngineSync->getLeaderChannel()->getEngineBuffer() != nullptr)) {
+        // Only set a new buffer size if we can determine it, otherwise we assume that it's not changed
+        m_currentLatency =
+                std::chrono::microseconds(m_pEngineSync->getLeaderChannel()
+                                                  ->getEngineBuffer()
+                                                  ->getAudioBufferMicros());
+    };
+}
+
+std::chrono::microseconds AbletonLink::getHostTime() const {
+    return m_hostTimeAtStartCallback + m_pLink->clock().micros() - m_timeAtStartCallback;
+}
+
+// Approximate the system time when the first sample in the current audio buffer will hit the speakers
+std::chrono::microseconds AbletonLink::getHostTimeAtSpeaker(
+        const std::chrono::microseconds hostTime) const {
+    return hostTime - m_currentLatency;
```

Comment:
I think we need to use here the time when the samples reach the DAC. 
See: VisualPlayPosition::setCallbackEntryToDacSecs